src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -498,46 +527,57 @@ def expectation_from_circuit(self, circuit: "Circuit", nshots: int = 1000) -> fl
+            # for each term that can be diagonalized simultaneously
+            # extract the coefficient, Z observable and qubit map
+            # the basis rotation, instead, will be the same
+            tmp_obs = []
+            measurements = {}
+            for term in terms:
+                # Only care about non-I terms
+                non_identity_factors = []
+                # prepare the measurement basis and append it to the circuit
+                for factor in term.factors:
+                    if factor.name[0] != "I":
+                        non_identity_factors.append(factor)
+                        q = factor.target_qubit
+                        if q not in measurements:
+                            measurements[q] = gates.M(q, basis=factor.gate.__class__)
+
+                # build diagonal observable
+                # including the coefficient
+                tmp_obs.append(
+                    SymbolicHamiltonian(
+                        term.coefficient
+                        * prod(
+                            Z(factor.target_qubit) for factor in non_identity_factors
+                        ),
+                        nqubits=circuit.nqubits,
+                        backend=self.backend,
+                    )
                 )
-            )
+
             # Get the qubits we want to measure for each term
-            qubit_map = sorted(factor.target_qubit for factor in non_identity_factors)
-            # prepare the measurement basis and append it to the circuit
-            measurements = [
-                gates.M(factor.target_qubit, basis=factor.gate.__class__)
-                for factor in non_identity_factors
-            ]
+            qubit_maps.append(sorted(measurements.keys()))
+            Z_observables.append(tmp_obs)
+
             circ_copy = circuit.copy(True)
-            circ_copy.add(measurements)
+            circ_copy.add(list(measurements.values()))
             rotated_circuits.append(circ_copy)
-            # for mapping the obtained sample frequencies to the original qubits
-            qubit_maps.append(qubit_map)
-        frequencies = [
-            result.frequencies()
-            for result in self.backend.execute_circuits(rotated_circuits, nshots=nshots)
-        ]
-        return sum(
-            coeff * obs.expectation_from_samples(freq, qubit_map)
-            for coeff, freq, obs, qubit_map in zip(
-                coefficients, frequencies, Z_observables, qubit_maps
-            )
-        )
+
+        # execute the circuits
+        results = self.backend.execute_circuits(rotated_circuits, nshots=nshots)
+
+        # construct the expectation value for each diagonal term
+        # and sum all together
+        expval = 0.0
+        for res, obs, qmap in zip(results, Z_observables, qubit_maps):
+            freq = res.frequencies()
+            expval += sum(o.expectation_from_samples(freq, qmap) for o in obs)
+        return expval
```

Comment:
```suggestion
        return self.constant + expval
```

Hmm, I think this is an existing bug, the constant term wasn't being added anywhere else?